From 3b45f6480c487ce7a387dc7a862a1097dc755be2 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Fri, 1 Sep 2023 15:47:35 +0200
Subject: [PATCH 063/120] gpio: stm32_gpio: release RIF semaphores before OS
 boot

GPIO is a RIF-aware IP, meaning it has to handle its RIF configuration
by its own.
The RIF semaphore can be acquired either with 'gpio' command or
through the pinctrl's set-state.

The OS in the following boot stage could require to allocate a shared
GPIO to another CPU, thus the semaphores should be released before
the OS boot.

For the 'gpio' command, release the semaphore when the GPIO is
released.
For pinctrl, release all the acquired semaphores during OS prepare.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 2ba2583135af33924f52eb01d15e06233b0c34a1 gpio: stm32_gpio: release RIF semaphores before OS boot]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I63b4f83f8d12f4f3cc14ef34d58357fdcc0b12d7
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/325460
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/gpio/stm32_gpio.c | 56 ++++++++++++++++++++++++++++++++++++++-
 1 file changed, 55 insertions(+), 1 deletion(-)

diff --git a/drivers/gpio/stm32_gpio.c b/drivers/gpio/stm32_gpio.c
index 6a0d3b453ac..97d6e952f59 100644
--- a/drivers/gpio/stm32_gpio.c
+++ b/drivers/gpio/stm32_gpio.c
@@ -140,6 +140,19 @@ bool stm32_gpio_rif_valid(struct stm32_gpio_regs *regs, unsigned int offset)
 	return false;
 }
 
+static void stm32_gpio_rif_release(struct stm32_gpio_regs *regs, unsigned int offset)
+{
+	u32 cid;
+
+	cid = readl(&regs->rif[offset].cidcfgr);
+
+	if (!(cid & STM32_GPIO_CIDCFGR_CFEN))
+		return;
+
+	if (cid & STM32_GPIO_CIDCFGR_SEMEN && cid & STM32_GPIO_CIDCFGR_SEMWL_CID1)
+		writel(0, &regs->rif[offset].semcr);
+}
+
 static int stm32_gpio_request(struct udevice *dev, unsigned offset, const char *label)
 {
 	struct stm32_gpio_priv *priv = dev_get_priv(dev);
@@ -169,6 +182,18 @@ static int stm32_gpio_request(struct udevice *dev, unsigned offset, const char *
 	return 0;
 }
 
+static int stm32_gpio_rfree(struct udevice *dev, unsigned int offset)
+{
+	struct stm32_gpio_priv *priv = dev_get_priv(dev);
+	struct stm32_gpio_regs *regs = priv->regs;
+	ulong drv_data = dev_get_driver_data(dev);
+
+	if (drv_data & STM32_GPIO_FLAG_RIF_CTRL)
+		stm32_gpio_rif_release(regs, offset);
+
+	return 0;
+}
+
 static int stm32_gpio_direction_input(struct udevice *dev, unsigned offset)
 {
 	struct stm32_gpio_priv *priv = dev_get_priv(dev);
@@ -330,6 +355,7 @@ static int stm32_gpio_get_flags(struct udevice *dev, unsigned int offset,
 
 static const struct dm_gpio_ops gpio_stm32_ops = {
 	.request		= stm32_gpio_request,
+	.rfree			= stm32_gpio_rfree,
 	.direction_input	= stm32_gpio_direction_input,
 	.direction_output	= stm32_gpio_direction_output,
 	.get_value		= stm32_gpio_get_value,
@@ -400,11 +426,39 @@ static int gpio_stm32_probe(struct udevice *dev)
 	return 0;
 }
 
+static int gpio_stm32_remove(struct udevice *dev)
+{
+	struct stm32_gpio_priv *priv = dev_get_priv(dev);
+	struct stm32_gpio_regs *regs = priv->regs;
+	ulong drv_data = dev_get_driver_data(dev);
+	unsigned int offset;
+	u32 seccfgr = 0;
+
+	if (!(drv_data & STM32_GPIO_FLAG_RIF_CTRL))
+		return 0;
+
+	if (drv_data & STM32_GPIO_FLAG_SEC_CTRL)
+		seccfgr = readl(&regs->seccfgr);
+
+	for (offset = 0; offset < STM32_GPIOS_PER_BANK; offset++) {
+		if (!stm32_gpio_is_mapped(dev, offset))
+			continue;
+
+		if ((seccfgr >> SECCFG_BITS(offset)) & SECCFG_MSK)
+			continue;
+
+		stm32_gpio_rif_release(regs, offset);
+	}
+
+	return 0;
+}
+
 U_BOOT_DRIVER(gpio_stm32) = {
 	.name	= "gpio_stm32",
 	.id	= UCLASS_GPIO,
 	.probe	= gpio_stm32_probe,
+	.remove	= gpio_stm32_remove,
 	.ops	= &gpio_stm32_ops,
-	.flags	= DM_UC_FLAG_SEQ_ALIAS,
+	.flags	= DM_UC_FLAG_SEQ_ALIAS | DM_FLAG_OS_PREPARE,
 	.priv_auto	= sizeof(struct stm32_gpio_priv),
 };
-- 
2.47.2

